### PR TITLE
Minor fix on sync insertion

### PR DESCRIPTION
**Pull request by shmsong**
Current raw sync insertion fails the following pattern (repro added):

``` cpp
__shared__ a, b;

a = ...;
// <- sync insertion point before fix.
for ...
  b = ...;
  // <- correct sync insertion in this case. 
  for ...
    ... = ... + a;
   ... = ...  + b;

````

Mechanically what happened is that the raw sync insertion pass decided to insert sync when seeing `    ... = ... + a;`, applies a sync at the loopnest level of a's producer `a=...`) and wipes clean all the smem tracking content, in this case both (`a=...` and `b=...`). 

Fix is essentially records all the smem writes that were handled by the inserted sync instead of only the producers of the shared mem tensor, so that the sync insertion point can be decided taking all the previous writes into consideration.

### CODE REVIEW

**Comment by naoyam** [comment]
Is the correct place before the write to `b`? Shouldn't it be after the write?

**Comment by shmsong** [comment]
> Is the correct place before the write to `b`? Shouldn't it be after the write?

Ah, yes. A typo on description. Thanks.

**Comment by naoyam** [comment]
Running the test on my system resulted in a capacity error with shmem. I'm using a Titan RTX. Can you please decrease the memory usage?

**Comment by naoyam** [comment]
Here's the error:

```
C++ exception with description "(dynamic_smem_size) < (available_dynamic_smem_without_reconfiguration + additional_dynamic_smem_available_through_reconfiguration) INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/executor.cpp":571, please report a bug to PyTorch. The total shared memory allocation is larger than available memory. Dynamic size: 66560. Available size: 49136. Configured smem size: 49152. Device limit size: 65536
Exception raised from computeLaunchParams at ../torch/csrc/jit/codegen/cuda/executor.cpp:571 (most recent call first):
```

**Comment by shmsong** [comment]
> Here's the error:
> 
> ```
> C++ exception with description "(dynamic_smem_size) < (available_dynamic_smem_without_reconfiguration + additional_dynamic_smem_available_through_reconfiguration) INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/executor.cpp":571, please report a bug to PyTorch. The total shared memory allocation is larger than available memory. Dynamic size: 66560. Available size: 49136. Configured smem size: 49152. Device limit size: 65536
> Exception raised from computeLaunchParams at ../torch/csrc/jit/codegen/cuda/executor.cpp:571 (most recent call first):
> ```

Sure I can reduce it to 40KB. Larger size makes it fail all the time when it's not sync'ed. 40KB should be good for all devices.

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp [hunk 1]
```diff
@@ -22262,6 +22262,48 @@ TEST_F(NVFuserTest, FusionRAWSyncInsertionPlace3_CUDA) {
   testValidate(&fusion, cg_outputs, {t0, t1}, {ref}, __LINE__, __FILE__);
 }
 
+TEST_F(NVFuserTest, FusionRAWSyncInsertionPlace4_CUDA) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu.cpp:22266]: Please add a comment here referencing this PR.

**Review by naoyam — APPROVED** [review]
The fix makes sense. Thanks. Please update the test so that it doesn't fail due to the memory limit.

**Comment by naoyam** [comment]
> Sure I can reduce it to 40KB. Larger size makes it fail all the time when it's not sync'ed. 40KB should be good for all devices.

Ideally, rather than relying on reproducing a result validation error, the test should check if the lowered kernel has a sync at the correct location.

**Pull request merged**